src/compile-cache.js

```diff
@@ -0,0 +1,131 @@
+'use strict'
+
+var path = require('path')
+var CSON = require('season')
+var fs = require('fs-plus')
+var _ = require('underscore-plus')
+
+var COMPILERS = {
+  '.js': require('./babel'),
+  '.ts': require('./typescript'),
+  '.coffee': require('./coffee-script')
+}
+
+var cacheDirectory = null
+
+_.each(COMPILERS, function (compiler, extension) {
+  Object.defineProperty(require.extensions, extension, {
+    enumerable: true,
+    writable: false,
+    value: function (module, filePath) {
+      var code = compileFileAtPath(compiler, filePath, extension)
+      return module._compile(code, filePath)
+    }
+  })
+})
+
+exports.setAtomHomeDirectory = function (atomHome) {
+  var cacheDir = path.join(atomHome, 'compile-cache')
+  if (process.env.USER === 'root' && process.env.SUDO_USER && process.env.SUDO_USER !== process.env.USER) {
+    cacheDir = path.join(cacheDirectory, 'root')
+  }
+  this.setCacheDirectory(cacheDir)
+}
+
+exports.setCacheDirectory = function (directory) {
+  cacheDirectory = directory
+  CSON.setCacheDir(path.join(cacheDirectory, 'cson'));
+}
+
+exports.getCacheDirectory = function () {
+  return cacheDirectory
+}
+
+exports.addPathToCache = function (filePath, atomHome) {
+  this.setAtomHomeDirectory(atomHome)
+  var extension = path.extname(filePath)
+
+  if (extension === '.cson') {
+    CSON.readFileSync(filePath)
+  } else {
+    var compiler = COMPILERS[extension]
```

Comment:
We had to change the code for typescript, coffee and babel to make source maps work properly. It would have been nice to refactor the CSON code to use the same pattern, but there wasn't a pressing need to touch `season`, so we left it as is for now.